main.c

```diff
@@ -12,13 +28,96 @@ typedef enum
 
 typedef struct config_t
 {
+  run_mode_t run_mode;
   brute_mode_t brute_mode;
   int length;
   char *alph;
   char *hash;
 } config_t;
 
-typedef bool (*password_handler_t) (char *password, void *context);
+typedef struct task_t
+{
+  password_t password;
+} task_t;
+
+typedef struct my_sem_t
+{
+  pthread_cond_t cond_sem;
+  pthread_mutex_t mutex;
+  int counter;
+} my_sem_t;
+
+void
+my_sem_init (my_sem_t *sem, int pshared, unsigned int value)
+{
+  (void)pshared; /* to suppress the "unused parameter" warning */
+  sem->counter = value;
+  pthread_cond_init (&sem->cond_sem, NULL);
+  pthread_mutex_init (&sem->mutex, NULL);
+}
+
+void
+my_sem_post (my_sem_t *sem)
+{
+  pthread_mutex_lock (&sem->mutex);
+  ++sem->counter;
+  pthread_cond_signal (&sem->cond_sem);
```

Comment:
Signalling is an expensive kernel call. It make sense to make it only if counter was 0 before increment.